File: scripts/analyze_sample_counts.R
#!/usr/bin/env Rscript

# Analyze sample counts in extraction files
library(dplyr)
library(purrr)

source("R/core/extraction_data_utils.R")

# Get all extraction files
extraction_dir <- "data/extractions"
files <- list.files(extraction_dir, pattern = "\\.xlsx$", full.names = TRUE)

cat(sprintf("Found %d extraction files\n\n", length(files)))

# Analyze each file
file_analysis <- map_dfr(files, function(file) {
  tryCatch({
    df <- readxl::read_excel(file)

    # Count total rows
    total_rows <- nrow(df)

    # Count non-empty rows (rows with at least some data)
    non_empty_rows <- sum(apply(df, 1, function(row) any(!is.na(row) & row != "")))

    # Try to find barcode column
    barcode_cols <- grep("barcode|code.?barre", names(df), ignore.case = TRUE, value = TRUE)
    has_barcodes <- length(barcode_cols) > 0

    barcode_count <- 0
    if (has_barcodes) {
      barcode_col <- barcode_cols[1]
      barcodes <- df[[barcode_col]]
      barcode_count <- sum(!is.na(barcodes) & barcodes != "")
    }

    data.frame(
      file = basename(file),
      total_rows = total_rows,
      non_empty_rows = non_empty_rows,
      has_barcodes = has_barcodes,
      barcode_count = barcode_count
    )
  }, error = function(e) {
    data.frame(
      file = basename(file),
      total_rows = NA,
      non_empty_rows = NA,
      has_barcodes = FALSE,
      barcode_count = 0,
      error = as.character(e$message)
    )
  })
})

# Print summary
cat("File-by-file analysis:\n")
cat("======================\n\n")
print(file_analysis %>% arrange(desc(total_rows)))

cat("\n\nSummary Statistics:\n")
cat("===================\n")
cat(sprintf("Total files: %d\n", nrow(file_analysis)))
cat(sprintf("Files with barcodes: %d\n", sum(file_analysis$has_barcodes)))
cat(sprintf("Total rows across all files: %d\n", sum(file_analysis$total_rows, na.rm = TRUE)))
cat(sprintf("Total non-empty rows: %d\n", sum(file_analysis$non_empty_rows, na.rm = TRUE)))
cat(sprintf("Total barcodes found: %d\n", sum(file_analysis$barcode_count, na.rm = TRUE)))
cat(sprintf("Average rows per file: %.1f\n", mean(file_analysis$total_rows, na.rm = TRUE)))
cat(sprintf("Median rows per file: %.1f\n", median(file_analysis$total_rows, na.rm = TRUE)))

# Now load using the actual app logic and compare
cat("\n\nLoading data using app logic:\n")
cat("==============================\n")
df_app <- load_all_extractions(extraction_dir)
metrics <- summarise_extraction_metrics(df_app)

cat(sprintf("Total samples (after deduplication): %d\n", metrics$total))
cat(sprintf("Files with barcodes: %d\n", metrics$files_with_barcodes))

# Check for duplicates
if ("source_file" %in% names(df_app)) {
  cat("\n\nSamples per source file:\n")
  cat("========================\n")
  samples_per_file <- df_app %>%
    group_by(source_file) %>%
    summarise(count = n()) %>%
    arrange(desc(count))
  print(samples_per_file)
}

File: scripts/analyze_sample_counts.py
#!/usr/bin/env python3

import pandas as pd
import glob
import os
from pathlib import Path

# Get all extraction files
extraction_dir = "data/extractions"
files = sorted(glob.glob(os.path.join(extraction_dir, "*.xlsx")))

print(f"Found {len(files)} extraction files\n")

file_analysis = []

# Analyze each file
for file_path in files:
    try:
        # Read the Excel file
        df = pd.read_excel(file_path)

        total_rows = len(df)

        # Count non-empty rows (rows with at least one non-null value)
        non_empty_rows = df.dropna(how='all').shape[0]

        # Look for barcode columns
        barcode_cols = [col for col in df.columns if 'barcode' in str(col).lower() or 'code' in str(col).lower()]
        has_barcodes = len(barcode_cols) > 0

        barcode_count = 0
        if has_barcodes and len(barcode_cols) > 0:
            barcode_col = barcode_cols[0]
            barcode_count = df[barcode_col].notna().sum()

        file_analysis.append({
            'file': os.path.basename(file_path),
            'total_rows': total_rows,
            'non_empty_rows': non_empty_rows,
            'has_barcodes': has_barcodes,
            'barcode_count': barcode_count
        })
    except Exception as e:
        file_analysis.append({
            'file': os.path.basename(file_path),
            'total_rows': 0,
            'non_empty_rows': 0,
            'has_barcodes': False,
            'barcode_count': 0,
            'error': str(e)
        })

# Create DataFrame for analysis
analysis_df = pd.DataFrame(file_analysis)

# Print file-by-file analysis
print("File-by-file analysis:")
print("=" * 80)
pd.set_option('display.max_rows', None)
pd.set_option('display.width', 200)
print(analysis_df.sort_values('total_rows', ascending=False).to_string(index=False))

# Print summary statistics
print("\n\nSummary Statistics:")
print("=" * 80)
print(f"Total files: {len(analysis_df)}")
print(f"Files with barcodes: {analysis_df['has_barcodes'].sum()}")
print(f"Total rows across all files: {analysis_df['total_rows'].sum()}")
print(f"Total non-empty rows: {analysis_df['non_empty_rows'].sum()}")
print(f"Total barcodes found: {analysis_df['barcode_count'].sum()}")
print(f"Average rows per file: {analysis_df['total_rows'].mean():.1f}")
print(f"Median rows per file: {analysis_df['total_rows'].median():.1f}")
print(f"Min rows: {analysis_df['total_rows'].min()}")
print(f"Max rows: {analysis_df['total_rows'].max()}")

# Show some sample data from the first file with barcodes
print("\n\nSample data from first file:")
print("=" * 80)
if len(files) > 0:
    sample_file = files[0]
    print(f"File: {os.path.basename(sample_file)}")
    df_sample = pd.read_excel(sample_file)
    print(f"\nColumns: {', '.join(df_sample.columns.tolist())}")
    print(f"\nFirst 5 rows:")
    print(df_sample.head().to_string())
    print(f"\nShape: {df_sample.shape[0]} rows × {df_sample.shape[1]} columns")
